### DISCUSSION TITLE
v2 Roadmap

### DISCUSSION THREAD

**Discussion by Quantalabs** [Quantalabs/DiffPriv#33] [Announcements/Releases]
DiffPriv v2 is going to be released in a couple months, and right now at the very beginning of this development process, it's best to lay out the v2 structure, roadmap, and plans for v1 migration.

## The Roadmap
Our Roadmap involves the release of several alpha, beta, and release candidates before releasing the initial v2 version, and other important milestone dates.

- **November 12, 2021** - Initial Alpha release of `v2a1`
   - Initial restructuring of the package
   - No pre-built Differential Privacy Mechanisms and Encryption Schemes, and no CLI (though all functions that existed previously are still there.)
- **December 5, 2021** - `v2a2` 
   - Adds pre-built Differential Privacy Mechanisms and Encryption Schemes
   - Additional features
- **December 26, 2021** - `v2a3`
  - Additional Differential Privacy Mechanisms and Encryption Schemes
  - Add manipulation and additional configuration of Differential Privacy Mechanisms and Encryption Schemes
- **December 28, 2021** - Initial Beta (bug and security fix only) `v2b1`
  - Bug fixes.
- **January 9, 2022** - `v2b2` 
  - Additional bug fixes (this release might not come if no bugs are found.)
- **January 30, 2022** - Initial release candidate - `v2rc1`, however, there could be a beta with additional bug fixes before this.
  - Sanity Checks
  - Security Checks
  - Improved documentation
- **April 9, 2022** - General Availability Release `v2.0.0`

<sup>1</sup> Additional pre-alpha releases might be released in-between alpha releases
<sup>2</sup> From the initial release candidate to the General Availability release, beta or other release-candidates might be published.
<sup>3</sup> The dates may change, and delays are not included. Do not rely on these release dates. 

## Migration

As this is a v2 release, there will be deprecations, and so migration is extremely important. We will include a migration guide in the documentation, and it will also be available at `/migration` on our [website](https://quantalabs.github.io/DiffPriv). This will stay there until we release a v3 - don't worry, this isn't coming anytime soon - at which point it will be moved to `/migration-v1-v2`, and the v2-v3 migration will take over at `/migration`. Remember, even if your release's support for security and bug fixes has been deprecated, we can still answer your questions and provide workarounds, just submit a discussion.

Even though v2 will be released, we have plans to keep on issuing patch and minor releases on v1, to help fix bugs and introduce new features for those who don't want to migrate, and will continue support for a while after v2's general availability release.

## Structure

With each release, DiffPriv has got increasingly complex, from one function in `v0.0.1`, to the CLI interface, encryption schemes, and privacy mechanisms in `v1.0.0`. However, a main goal of DiffPriv is to allow for endless customizability, in as many ways as we can, and currently, we don't feel like it is that customizable.

The restructuring will change the way DiffPriv is used, instead of individual functions and classes, everything revolves around the mechanism or the scheme, for differential privacy or encryption, respectively. So, DiffPriv's module's will look like this:
```
DiffPriv
  ├── scheme 
  │    └── pre
  │    └── utils
  ├── mech
  │     ├── pre
  │     ├── validate
  │     └── utils
  ├── cli
  └── utils
```
Where:
- `scheme` - Contains the scheme class,
  - `pre` - Contains pre-built encryption schemes
  - `utils` - Encryption related utilities.
- `mech` - Contains the mechanism class,
  - `pre` - Contains pre-built mechanisms
  - `validate` - Validation of mechanism parameters.
  - `utils` - Privacy related utilities.
- `utils` - Warnings, Errors, and other utilities for the package.
- `cli` - CLI interface

**Comment by quantum9Innovation** [comment]
Work has already started in branch https://github.com/Quantalabs/DiffPriv/tree/patch-33/. Merging into https://github.com/Quantalabs/DiffPriv/tree/v2 as soon as possible.

**Comment by Quantalabs** [comment]
Branch [`patch-34-enc-v2a1`](https://github.com/Quantalabs/DiffPriv/tree/patch-34-enc-v2a1) is working on the `mech` class.

**Reply by quantum9Innovation** [reply]
Already done in https://github.com/Quantalabs/DiffPriv/tree/patch-33/

**Reply by Quantalabs** [reply]
Whoops, I was referring to the `scheme` class 😆

**Comment by Quantalabs** [comment]
`v2a1` has been released (#36)! Next release we'll add prebuilt mechanisms, and scheme class, though the scheme class might also be postponed to the next release.

**Comment by Quantalabs** [comment]
`v2a2` has been released (#37)! We are postponing the scheme class and prebuilt schemes to `v2a3` and `a4`, respectively. This release does add however prebuilt mechanisms. Might take a bit to get published to conda.

**Comment by Quantalabs** [comment]
December 26 and 28th releases will be skipped. Initial beta will be published after another alpha release on January 3rd. The scheme class has been merged into the mech class, and the beta will fix any issues with this.
